### PR TITLE
fix(RHINENG-3405): limit API events sent to Segment

### PR DESCRIPTION
**Pull request by marleystipich2**
# Overview

This PR is being created to address [RHINENG-3405](https://issues.redhat.com/browse/RHINENG-3405).
We're currently sending an event for every Inventory API call whether it's from UI, CLI, or command line tool but the only events that we actually want to see are the requests done through a command line tool(ex. curl). This means we're sending millions of events to Segment that we are not actually using and are not providing any business value. The total monthly event limit for all Red Hat sources is 150 million per month and our Red Hat Insights API - Prod itself is sending about 2.5 million events per day. 

We need to filter these events in the code where we are doing the track call so that we are only sending the necessary events. The only events that we need to track in Segment are API requests that come from a command line tool. This should drop our event count from 2.5 million to about 260,000 events per day. We will also be able to remove our destination filters in Segment.

## PR Checklist

- [ ] Keep PR title short, ideally under 72 characters
- [ ] Descriptive comments provided in complex code blocks
- [ ] Tests: validate optimal/expected output
- [ ] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [ ] Uses [type hinting](https://docs.python.org/3/library/typing.html), if convenient
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: api/segmentio.py [hunk 1]
```diff
@@ -1,13 +1,18 @@
...
 def segmentio_track(op_name, processing_time, contextual_data, logger):
     if analytics.write_key:
         user_agent_header = connexion.request.headers.get("User-Agent", "")
-        if not user_agent_header:
+        exp = re.compile('Mozilla|Satellite|OpenAPI-Generator|insights-client')
+
+        # Do not send track event for API requests from web browser, CLI, QE tests
+        # or requests with missing User-Agent header
+        if not user_agent_header or bool(re.search(exp, user_agent_header)):
             logger.debug("Skipping analytics.track(), User-Agent header not set")
```
Comment (roliveri) [inline api/segmentio.py:16]: This debug statement will indicate that user_agent_header is not set even for cases when it is set. Maybe remove the reason in the message, or indicate which case is causing us to skip. The latter may be more useful when debugging in the future.

**Review by roliveri — CHANGES_REQUESTED** [review]

**Review by roliveri — APPROVED** [review]
LGTM

File: api/segmentio.py [hunk 2]
```diff
@@ -1,14 +1,20 @@
...
 def segmentio_track(op_name, processing_time, contextual_data, logger):
     if analytics.write_key:
         user_agent_header = connexion.request.headers.get("User-Agent", "")
-        if not user_agent_header:
-            logger.debug("Skipping analytics.track(), User-Agent header not set")
+
+        # Do not send track event for API requests from web browser, CLI, QE tests
+        # or requests with missing User-Agent header
+        if not user_agent_header or bool(re.search(USER_AGENT_IDENTIFIER, user_agent_header)):
+            logger.debug("Skipping analytics.track(), request came from web browser, CLI or \
```
[the remark this answers is not in the document]
Comment (marleystipich2) [reply api/segmentio.py:16]: We don't want to track requests that came from the UI, CLI or QE tests and the User-Agent is set for these requests so I don't think saying `request did not use user-agent` would make sense in this case. 

So basically we don't want to run the `analytics.track` call when the user agent contains one of these words `Mozilla|Satellite|OpenAPI-Generator|insights-client` or when the user-agent isn't set.

[the remark this answers is not in the document]
Comment (marleystipich2) [reply api/segmentio.py:16]: Not sure if this clears up the debug message or I can try to think of something that makes more sense

[the remark this answers is not in the document]
Comment (thearifismail) [reply api/segmentio.py:16]: I thought call trigged by CLI  were to be tracked.  The question is what are the sources/triggers to executed API calls besides `Mozilla|Satellite|OpenAPI-Generator|insights-client`?

**Review by thearifismail — COMMENTED** [review]
Approving the PR despite my questions because the questions are about a debug message.

**Review by thearifismail — APPROVED** [review]

File: api/segmentio.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (marleystipich2) [reply api/segmentio.py:16]: Sorry, by CLI I meant specifically API calls triggered from insights-client. We don't want to track these. The user agent will look something like this for insights client: `insights-client/3.1.8 (Core 3.2.23; requests 2.6.0) Red Hat Enterprise Linux Server 7.9 (CPython 2.7.5; Linux 3.10.0-1160.102.1.el7.x86_64); systemd`

We **do** want to track customers that are accessing the Inventory API 'directly' using a command line tool like `curl` for example. So these are the only API calls that we want to track.

**Comment by sthirugn** [comment]
any help to get this quickly to stage and prod is much appreciated!

**Review by FabriciaDinizRH — DISMISSED** [review]
The build errors are probably related to the quay outages, but we're holding off merging PRs right now (it might be ok to merge it later today though)

**Review by kruai — APPROVED** [review]

**Comment by kruai** [comment]
@marleystipich2 Please use [pre-commit](https://github.com/RedHatInsights/insights-host-inventory#contributing) when contributing to the repo; otherwise, the PR checks will fail.

**Comment by kruai** [comment]
Based on @sthirugn's comment and Frank pinging me this morning, I'm assuming this is urgent, so I've pushed a fix. If the PR checks don't catch anything nasty then I'll merge it

**Comment by marleystipich2** [comment]
@kruai Good to know about pre-commit. Thanks for the help!

**Comment by kruai** [comment]
There are some QE test failures, but they're the same ones that were occurring in #1516; my understanding from that PR is that those are expected failures and they're being worked on (i.e. not related to this PR's changes)

**Pull request merged**